File: ain5/spko/funktional/funktional/funktional.cpp
// funktional.cpp
#include <iostream>
#include <iomanip>
#include <functional>

namespace {
    double einlesen() {
        double d;
        std::cin >> d;
        return d;
    }

    // Funktion hoeherer Ordung: Funktion als Rueckgabewert
    std::function<double(double)> steuer_berechnung(bool ermaessigt) {
        double steuer = ermaessigt ? 1.07 : 1.19;
        return [=](double netto){ return netto * steuer; };
    }

    // Funktion hoeherer Ordung: Funktionen als Parameter
    double brutto(std::function<double()> netto, std::function<double(double)> versteuern) {
        return versteuern(netto());
    }
}

int main() {
    std::cout << std::fixed << std::setprecision(2)
              << brutto(einlesen, steuer_berechnung(true)) << '\n';
    std::cout << std::fixed << std::setprecision(2)
              << brutto(einlesen, steuer_berechnung(false)) << '\n';
}

File: ain5/spko/funktional/funktional/Funktional.java
// Funktional.java
import java.util.Scanner;
import java.util.function.UnaryOperator;
import java.util.function.DoubleSupplier;

public final class Funktional {
    private Funktional() { }

    private static final Scanner EINGABE = new Scanner(System.in);

    public static void main(String[] args) {
        System.out.printf(
            "%.2f%n", brutto(EINGABE::nextDouble, steuerBerechnung(true)));
        System.out.printf(
            "%.2f%n", brutto(EINGABE::nextDouble, steuerBerechnung(false)));
    }

    // Funktion hoeherer Ordung: Funktion als Rueckgabewert
    private static
    UnaryOperator<Double> steuerBerechnung(boolean ermaessigt) {
        double steuer = ermaessigt ? 1.07 : 1.19;
        return (netto) -> netto * steuer;
    }

    // Funktion hoeherer Ordung: Funktionen als Parameter
    private static
    double brutto(DoubleSupplier netto, UnaryOperator<Double> versteuern) {
        return versteuern.apply(netto.getAsDouble());
    }
}

